### PR TITLE
i#1815: fix unit_test failure on macos sierra due to gettimeofday

### PR DESCRIPTION
**Pull request by shawndenbow**
On MacOS before Sierra, gettimeofday returns usecs:secs and does not
set the timeval struct. Sierra correctly fills the timeval struct.

### CODE REVIEW

File: core/unix/os.c [hunk 1]
```diff
@@ -347,6 +347,16 @@ static byte *app_brk_cur;
...
 #endif
 
+#ifdef MACOS
+static int macos_version;
+# define MACOS_VERSION_SIERRA 16
+# define MACOS_VERSION_EL_CAPITAN 15
+# define MACOS_VERSION_YOSEMITE 14
+# define MACOS_VERSION_MAVERICKS 13
+# define MACOS_VERSION_MOUNTAIN_LION 12
+# define MACOS_VERSION_LION 11
```
Comment (derekbruening) [inline core/unix/os.c]: (Looks good static in this .c for now, but in the future we might want to consider exposing these via the dr_get_os_version() API.)

File: core/unix/os.c [hunk 2]
```diff
@@ -1135,39 +1146,43 @@ get_timer_frequency()
...
-#else
-    /* SYS_time is considered obsolete, and is unsupported on ARM */
     struct timeval current_time;
-    if (dynamorio_syscall(SYS_gettimeofday, 2, &current_time, NULL) >= 0) {
+    uint64 val = dynamorio_syscall(SYS_gettimeofday, 2, &current_time, NULL);
+#ifdef MACOS
+    /* MacOS before Sierra returns usecs:secs and does not set the timeval struct. */
+    if (macos_version < MACOS_VERSION_SIERRA) {
+        if ((int)val < 0)
+          return 0;
```
Comment (derekbruening) [inline core/unix/os.c]: style: indentation off

File: core/unix/os.c [hunk 3]
```diff
@@ -1135,39 +1146,43 @@ get_timer_frequency()
...
-    if ((int)val > 0) {
-#else
-    if (dynamorio_syscall(SYS_gettimeofday, 2, &current_time, NULL) >= 0) {
+#ifdef MACOS
+    /* MacOS before Sierra returns usecs:secs and does not set the timeval struct. */
+    if (macos_version < MACOS_VERSION_SIERRA) {
+        if ((int)val > 0) {
+            current_time.tv_sec = (uint) val;
+            current_time.tv_usec = (uint)(val >> 32);
+        } else {
```
Comment (derekbruening) [inline core/unix/os.c]: This else could be removed as we'll hit the else down below?

File: core/unix/os.c [hunk 4]
```diff
@@ -1183,15 +1198,20 @@ uint64
...
```
Comment (derekbruening) [inline core/unix/os.c]: This else could also be removed?

File: core/unix/os.c (hunk 1 above)
Comment (shawndenbow) [inline core/unix/os.c]: Agree. I added in a comment to xref i#1404.

File: core/unix/os.c (hunk 2 above)
Comment (shawndenbow) [inline core/unix/os.c]: Done.

File: core/unix/os.c (hunk 3 above)
Comment (shawndenbow) [inline core/unix/os.c]: Done.

File: core/unix/os.c (hunk 4 above)
Comment (shawndenbow) [inline core/unix/os.c]: Done.

**Pull request merged**